File: 백준/Silver/9742. 순열/순열.cc
#include <iostream>
using namespace std;

int N, length;
char str[11];
char result[11];
char answer[11];

int perm_count;
int dp[11];
int collected[11];

static void permutation(int depth) {
    if (depth == length) {
        if (N == perm_count) {
            for (int i = 0; i < length; i++)
            {
                answer[i] = result[i];
            }
        }

        perm_count++;

        return;
    }

    for (int i = 0; i < length; i++)
    {
        if (collected[i] == 0) {
            collected[i] = 1;
            result[depth] = str[i];
            permutation(depth + 1);
            collected[i] = 0;
        }
    }
}

int main() {
    cin.tie(0);
    cout.tie(0);
    ios::sync_with_stdio(false);

    dp[1] = 1;

    for (int i = 2; i < 11; i++)
    {
        dp[i] = dp[i - 1] * i;
    }

    while (true)
    {
        cin >> str >> N;

        perm_count = 1;

        for (int i = 0; i < 11; i++)
        {
            length = i;
            if (str[i] == NULL) break;
        }

        if (cin.eof()) break;

        if (dp[length] < N) {
            for (int i = 0; i < length; i++)
            {
                cout << str[i];
            }
            cout << " " << N << " = ";
            cout << "No permutation" << "\n";
        }
        else {
            permutation(0);
            for (int i = 0; i < length; i++)
            {
                cout << str[i];
            }
            cout << " " << N << " = ";
            for (int i = 0; i < length; i++)
            {
                cout << answer[i];
            }
            cout << "\n";
        }
    }

    return 0;
}

File: Cpp_Algo/백준/Silver/9742. 순열/순열.py
import sys

input = sys.stdin.readline


# Function to calculate factorial
def factorial(n):
    if n == 0 or n == 1:
        return 1
    return n * factorial(n - 1)


# Function to perform permutation
def permutation(depth):
    global perm_count
    global answer

    if depth == length:
        if N == perm_count:
            answer = result[:]
        perm_count += 1
        return

    for i in range(length):
        if not collected[i]:
            collected[i] = 1
            result[depth] = str[i]
            permutation(depth + 1)
            collected[i] = 0


# Initialize variables
dp = [0] * 11
dp[1] = 1

for i in range(2, 11):
    dp[i] = dp[i - 1] * i

# Main loop
while True:
    # Input
    try:
        line = input()

        if not line:
            break

        str, N = line.split()
        N = int(N)

        perm_count = 1

        length = len(str)
        str = list(str)
        result = [0] * length
        collected = [0] * length

        if dp[length] < N:
            print(''.join(str), N, "= No permutation")
        else:
            permutation(0)
            print(''.join(str), N, '=', ''.join(answer))
    except:
        break